### PR TITLE
VM->Broker Conversion for api/test_remoteexecution.py

### PR DESCRIPTION
**Pull request by JacobCallahan**
Converted the one test in this module that used libvirt-based
provisioning
Enhanced the Host, Capsule, and Satellite classes
Additionally, added a new class to helpers that helps when dealing with
installer command construction.

### CODE REVIEW

**Comment by JacobCallahan** [comment]
@lhellebr the test is currently failing at the last step, which I don't believe is a result of the conversion since all previous steps passed. I'm seeing a 500 error:
```requests.exceptions.HTTPError: 500 Server Error: Internal Server Error for url: https://satellite:443/api/v2/smart_proxies/5/refresh```

File: robottelo/helpers.py [hunk 1]
```diff
@@ -716,3 +716,64 @@ def idgen(val):
     test name
     """
     return '_parameter'
+
+
+class InstallerCommand:
+    """This class constructs, parses, updates and gets formatted installer commands"""
+
+    def __init__(self, *args, command='satellite-installer', allow_dupes=False, **kwargs):
```
Comment (tstrych) [inline robottelo/helpers.py:724]: Parsing methods are hell to debug, can you please add/provide as much documentation as possible.

Comment (tstrych) [inline robottelo/helpers.py:724]: allow_dupes - is allow duplicates I suppose :), why not use full naming ?

**Review by tstrych — APPROVED** [review]
Overall looks good. Of course I didn't went deeply through `class InstallerCommand`.
Please document methods and try to resolve the error you mentioned.

File: robottelo/hosts.py [hunk 2]
```diff
@@ -446,15 +462,57 @@ def check_services(self):
...
-        command_args.update(cmd_kwargs)
-        command_args = ' '.join(
-            [f'--{key.replace("_", "-")} {value}' for key, value in command_args.items()]
+        if not installer_obj:
+            command_opts = {'scenario': self.__class__.__name__.lower()}
+            command_opts.update(cmd_kwargs)
+            installer_obj = InstallerCommand(*cmd_args ** command_opts)
+        return self.execute(installer_obj.get_command())
+
+    def capsule_setup(self, sat_host, **installer_kwargs):
```
[the remark this answers is not in the document]
Comment (tstrych) [reply robottelo/hosts.py:473]: I guess other changes will require modification in the existing methods but that's okay. :+1:

[the remark this answers is not in the document]
Comment (JacobCallahan) [reply robottelo/hosts.py:473]: Correct, these are standard tasks for capsule setup. Test authors can moderately deviate within a test and still use this.

For major deviations, they can go entirely manual, the run the installer method.

File: robottelo/helpers.py (hunk 1 above)
Comment (JacobCallahan) [reply robottelo/helpers.py:724]: Variable names should be long enough to infer meaning, but not overly verbose.

**Comment by JacobCallahan** [comment]
> Overall looks good. Of course I didn't went deeply through `class InstallerCommand`.
> Please document methods and try to resolve the error you mentioned. 
> 

If Lukas has an idea for what caused the error and/or how to fix it, then I can include that. However, this is not a blocking requirement for this pull request.

**Comment by lhellebr** [comment]
From just ISE 500, I have no idea. I'd need to try manually and watch logs.

File: robottelo/helpers.py [hunk 3]
```diff
@@ -716,3 +716,64 @@ def idgen(val):
...
+                for v in val:
+                    command_str += f' --{key.replace("_", "-")} {v}'
+            else:
+                command_str += f' --{key.replace("_", "-")} {val}'
+        return command_str
+
+    def update(self, *args, allow_dupes=False, **kwargs):
+        new_args = [arg for arg in args if arg not in self.args]
+        self.args.extend(new_args)
+        if not allow_dupes:
```
Comment (omkarkhatavkar) [inline robottelo/helpers.py:745]: is this `if not allow_dupes:` is redundant ? or necessary two places where it is working same object similarly

File: robottelo/helpers.py [hunk 4]
```diff
@@ -716,3 +716,64 @@ def idgen(val):
...
+                self.opts[key] = val
+            elif (curr_val := self.opts.get(key)) :  # noqa: E203
+                val = [val]
+                if not isinstance(curr_val, list):
+                    curr_val = [curr_val]
+                val += curr_val
+            self.opts[key] = val
+
+    @classmethod
+    def from_cmd_str(cls, command='satellite-installer', cmd_str=None):
```
Comment (omkarkhatavkar) [inline robottelo/helpers.py:758]: is this only used to for certain scenarios ? can be used anywhere. can you add example how can this used in doc string something.

File: robottelo/hosts.py [hunk 5]
```diff
@@ -446,15 +462,57 @@ def check_services(self):
...
+    def install(self, installer_obj=None, cmd_args=None, cmd_kwargs=None):
         """General purpose installer"""
-        command_args = {'scenario': self.__class__.__name__.lower()}
-        command_args.update(cmd_kwargs)
-        command_args = ' '.join(
-            [f'--{key.replace("_", "-")} {value}' for key, value in command_args.items()]
+        if not installer_obj:
+            command_opts = {'scenario': self.__class__.__name__.lower()}
+            command_opts.update(cmd_kwargs)
+            installer_obj = InstallerCommand(*cmd_args ** command_opts)
```
Comment (omkarkhatavkar) [inline robottelo/hosts.py:470]: `InstallerCommand(*cmd_args ** command_opts)` missing ',' or I dont know this pattern

File: tests/foreman/api/test_remoteexecution.py [hunk 6]
```diff
@@ -24,15 +24,14 @@
 from robottelo.api.utils import wait_for_tasks
 from robottelo.config import settings
 from robottelo.helpers import add_remote_execution_ssh_key
-from robottelo.vm_capsule import CapsuleVirtualMachine
+from robottelo.hosts import Satellite
 
 
 CAPSULE_TARGET_VERSION = '6.9.z'
```
Comment (omkarkhatavkar) [inline tests/foreman/api/test_remoteexecution.py:30]: why it is hard-coded ? not part of this PR change

Comment (vsedmik) [reply tests/foreman/api/test_remoteexecution.py:30]: Used before, we could change to `CAPSULE_TARGET_VERSION = f'{settings.satellite_version}/z'`.

**Comment by vsedmik** [comment]
At 6.9 the test is passing locally with this PR. At 6.10 it fails due to missing 6.10.z scenario in foreman-maintain which shall be picked in one of next snaps with `rubygem-foreman_maintain-0.8.1`.
The capsule deployment and registration looks good. :+1:

File: robottelo/helpers.py (hunk 3 above)
Comment (JacobCallahan) [reply robottelo/helpers.py:745]: allow_dupes is here so people can either add an additional value for the same option or override an existing one.

File: robottelo/helpers.py (hunk 4 above)
Comment (JacobCallahan) [reply robottelo/helpers.py:758]: it can theoretically be used for any scenario/installer. well technically any command..

File: robottelo/hosts.py (hunk 5 above)
Comment (JacobCallahan) [reply robottelo/hosts.py:470]: ooohhhh good catch

File: robottelo/helpers.py [hunk 7]
```diff
@@ -716,3 +716,82 @@ def idgen(val):
...
+
+    def update(self, *args, allow_dupes=False, **kwargs):
+        """Update one or more arguments and options
+        values passed as positional and keyword arguments
+        """
+        new_args = [arg for arg in args if arg not in self.args]
+        self.args.extend(new_args)
+        if not allow_dupes:
+            self.opts.update(kwargs)
+        for key, val in kwargs.items():
```
Comment (mshriver) [inline robottelo/helpers.py:761]: I think this warrants some documentation. Slick little loop though, nice.

Comment (JacobCallahan) [reply robottelo/helpers.py:761]: anything you'd specifically like documented here?

Comment (mshriver) [reply robottelo/helpers.py:761]: nothing specifically, just a plain language description of the intent/action of the loop so its quicker to 'read'.

File: tests/foreman/api/test_remoteexecution.py [hunk 8]
```diff
@@ -24,15 +24,14 @@
...
 
 
 @pytest.mark.tier4
-@pytest.mark.libvirt_content_host
```
Comment (mshriver) [inline tests/foreman/api/test_remoteexecution.py:34]: :tada:

**Review by mshriver — APPROVED** [review]
